fendermint/app/src/app.rs

```diff
@@ -620,11 +623,42 @@ where
         *guard = Some(state);
 
         let response = match result {
-            Err(e) => invalid_check_tx(AppError::InvalidEncoding, e.description),
+            Err(e) => {
+                emit(MpoolReceivedInvalidMessage {
+                    reason: "InvalidEncoding",
+                    description: e.description.as_ref(),
+                });
+                invalid_check_tx(AppError::InvalidEncoding, e.description)
+            }
             Ok(result) => match result {
-                Err(IllegalMessage) => invalid_check_tx(AppError::IllegalMessage, "".to_owned()),
-                Ok(Err(InvalidSignature(d))) => invalid_check_tx(AppError::InvalidSignature, d),
-                Ok(Ok(ret)) => to_check_tx(ret),
+                Err(IllegalMessage) => {
+                    emit(MpoolReceivedInvalidMessage {
+                        reason: "IllegalMessage",
+                        description: "",
+                    });
+
+                    invalid_check_tx(AppError::IllegalMessage, "".to_owned())
+                }
+                Ok(Err(InvalidSignature(d))) => {
+                    emit(MpoolReceivedInvalidMessage {
+                        reason: "InvalidSignature",
+                        description: d.as_ref(),
+                    });
+                    invalid_check_tx(AppError::InvalidSignature, d)
+                }
+                Ok(Ok(ret)) => {
+                    emit(MpoolReceived {
+                        from: ret.message.from.to_string().as_str(),
+                        to: ret.message.to.to_string().as_str(),
+                        value: ret.message.value.to_string().as_str(),
+                        param_len: ret.message.params.len(),
+                        gas_limit: ret.message.gas_limit,
+                        fee_cap: ret.message.gas_fee_cap.to_string().as_str(),
+                        premium: ret.message.gas_premium.to_string().as_str(),
+                        accept: ret.exit_code.is_success(),
```

Comment:
It does indeed. I needed some of the parameters to be strings (for instance, for metrics labels) and wanted to have generic events. But you are right; it would be more efficient to just use the types directly (the ones that implement Display).

Reply:
Got it, can we attempt to preserve the types during emit, and push the conversion deeper in, to where it's actually needed (emitting the metric with labels)?

Reply:
Yes, I have already done it :)